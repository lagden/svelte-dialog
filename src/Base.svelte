<svelte:options accessors />

<script>
	export let isOpen = false

	/** @type {HTMLDialogElement} */
	let dialog
	let animation = false

	function animationend() {
		if (dialog) {
			animation = false
			isOpen = false
			dialog.close()
		}
	}

	export function open() {
		if (dialog && isOpen === false) {
			isOpen = true
			dialog.showModal()
		}
	}

	export function close() {
		if (dialog && isOpen === true) {
			animation = true
		}
	}

	/**
	 * @param {HTMLDialogElement} node
	 */
	function init(node) {
		globalThis.document.body.insertAdjacentElement('beforeend', node)
		node.addEventListener('animationend', animationend)
		return {
			destroy() {
				node.removeEventListener('animationend', animationend)
				node.remove()
			},
		}
	}
</script>

<dialog
	use:init
	bind:this={dialog}
	class:tadashi-svelte-dialog={true}
	class:tadashi-svelte-dialog--animation={animation}
	{...$$restProps}
><slot /></dialog>

<style>
	.tadashi-svelte-dialog[open] {
		opacity: 0;
		animation: dialog_in 0.3s ease-in 0.8s;
		animation-fill-mode: forwards;
	}

	.tadashi-svelte-dialog[open]::backdrop {
		animation: dialog_bg_in 0.3s ease-in;
	}

	.tadashi-svelte-dialog.tadashi-svelte-dialog--animation[open] {
		animation: dialog_out 0.3s ease-out;
	}

	.tadashi-svelte-dialog.tadashi-svelte-dialog--animation[open]::backdrop {
		animation: dialog_bg_out 0.3s ease-out;
		animation-fill-mode: forwards;
	}

	@keyframes dialog_bg_in {
		0% {
			opacity: 0;
		}
		100% {
			opacity: 1;
		}
	}

	@keyframes dialog_bg_out {
		0% {
			opacity: 1;
		}
		100% {
			opacity: 0;
		}
	}

	@keyframes dialog_in {
		0% {
			opacity: 0;
			transform: var(--tadashi-svelte-dialog-in-from-transform, scale(0.8));
		}
		100% {
			opacity: 1;
			transform: var(--tadashi-svelte-dialog-in-to-transform, scale(1));
		}
	}

	@keyframes dialog_out {
		0% {
			opacity: 1;
			transform: var(--tadashi-svelte-dialog-out-from-transform, scale(1));
		}
		100% {
			opacity: 0;
			transform: var(--tadashi-svelte-dialog-out-to-transform, scale(1.2));
		}
	}
</style>
