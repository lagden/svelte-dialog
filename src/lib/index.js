export * as acts from './acts.js'
export {default as Dialog} from './Dialog.svelte'
