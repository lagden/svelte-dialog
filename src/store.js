import {writable, get} from 'svelte/store'

const dialogMap = new Map()

const base = {
	open: false,
	title: '',
	message: '',
	component: {}
}

function toggle(n, v) {
	n.open = v
	return n
}

function dialogStore() {
	const {set, subscribe, update} = writable({...base})
	return {
		set,
		subscribe,
		open: () => update(n => toggle(n, true)),
		close: () => update(n => toggle(n, false))
	}
}

const dialog = dialogStore()

const unique = name => {
	if (dialogMap.has(name)) {
		return dialogMap.get(name)
	}
	dialogMap.set(name, dialogStore())
	return dialogMap.get(name)
}

const getUnique = name => {
	if (dialogMap.has(name)) {
		return dialogMap.get(name)
	}
}

const getData = store => get(store)

export {
	base,
	dialog,
	unique,
	getUnique,
	getData
}
