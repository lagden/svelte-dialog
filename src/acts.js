import * as store from './store'

function getDialog(name) {
	return name ? store.getUnique(name) : store.dialog
}

export function init(opts, name) {
	const dialog = name ? store.unique(name) : store.dialog
	opts = {
		...store.base,
		...opts
	}
	dialog.set(opts)
	return dialog
}

export function update(opts, name) {
	const dialog = getDialog(name)
	if (dialog) {
		opts = {
			...store.base,
			...opts
		}
		dialog.set(opts)
	}
}

export function open(name) {
	const dialog = getDialog(name)
	if (dialog) {
		dialog.open()
	}
}

export function close(name) {
	const dialog = getDialog(name)
	if (dialog) {
		dialog.close()
	}
}
