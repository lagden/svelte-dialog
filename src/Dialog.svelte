<script>
	import {onDestroy} from 'svelte'
	import {init} from './acts.js'

	import Base from './Base.svelte'

	export let name = undefined
	export let title = ''
	export let message = undefined
	export let component = undefined
	export let maxHeight = false
	export let useAlignTop = true
	export let showClose = true
	export let style = ''

	const opts = {
		title,
		message,
		component,
		maxHeight,
		useAlignTop,
		showClose,
	}

	const dialog = init(opts, name)

	function close() {
		$dialog.open = false
	}

	function handleKeydown(event) {
		if (event.key === 'Escape' && showClose) {
			event.preventDefault()
			close()
		}
	}

	const optsKeys = Object.keys(opts)
	const unsubscribe = dialog.subscribe(o => {
		for (const k of optsKeys) {
			opts[k] = o?.[k] ?? opts[k]
		}
	})

	onDestroy(() => {
		unsubscribe()
	})
</script>

<svelte:window on:keydown={handleKeydown}/>

{#if $dialog.open}
	<Base
		maxHeight={opts.maxHeight}
		useAlignTop={opts.useAlignTop}
		showClose={opts.showClose}
		on:click={close}
		{style}
	>
		<h3 class="_tadashi_svelte_dialog__title" slot="header">{opts.title}</h3>

		{#if opts.message}
			{@html opts.message}
		{/if}

		{#if opts.component}
			<svelte:component this={opts.component.element} {...opts.component?.props ?? {}} />
		{/if}

		<slot />
	</Base>
{/if}

<style>
:root {
	--tadashi_svelte_dialog__title_margin: 0;
	--tadashi_svelte_dialog__title_padding: 0;
	--tadashi_svelte_dialog__title_font_size: 1.5em;
}

._tadashi_svelte_dialog__title {
	margin: var(--tadashi_svelte_dialog__title_margin);
	padding: var(--tadashi_svelte_dialog__title_padding);
	font-size: var(--tadashi_svelte_dialog__title_font_size);
}
</style>
