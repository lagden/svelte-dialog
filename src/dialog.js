/* istanbul ignore file */

import * as acts from './acts.js'
import Dialog from './Dialog.svelte'

export {
	Dialog,
	acts,
}
