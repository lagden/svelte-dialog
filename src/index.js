/* istanbul ignore file */

import * as acts from './acts'
import Dialog from './Dialog.svelte'

export {
	Dialog,
	acts
}
