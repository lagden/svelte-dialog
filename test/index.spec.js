/* globals describe, afterEach, test, expect */

import {cleanup, render} from '@testing-library/svelte'
import {Dialog, acts} from '../src'

describe('Dialog', () => {
	afterEach(cleanup)

	test('should match snapshot', done => {
		const {container} = render(Dialog, {
			props: {
				name: 'teste',
				title: 'Teste',
				message: 'Apenas um show.'
			}
		})
		acts.open('teste')

		setTimeout(() => {
			expect(container).toMatchSnapshot()
			done()
		}, 1000)
	})
})
