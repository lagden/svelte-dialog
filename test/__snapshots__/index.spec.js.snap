// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should match snapshot 1`] = `
<body>
  <div>
    
  </div>
  <div
    class="_tadashi_svelte_dialog svelte-1x96u8l _tadashi_svelte_dialog__align_top"
    style=""
  >
    <div
      class="_tadashi_svelte_dialog__box svelte-1x96u8l"
    >
      <div
        class="_tadashi_svelte_dialog__content _tadashi_svelte_dialog__grid svelte-1x96u8l"
      >
        <div
          class="_tadashi_svelte_dialog__header svelte-1x96u8l"
        >
          <h3
            class="_tadashi_svelte_dialog__title svelte-1svxb87"
            slot="header"
          >
            Teste
          </h3>
        </div>
         
        <button
          class="_tadashi_svelte_dialog__close svelte-1x96u8l"
          type="button"
        >
          Ã—
        </button>
         
        <div
          class="_tadashi_svelte_dialog__body svelte-1x96u8l"
        >
          Apenas um show.
          
           
           
        </div>
      </div>
    </div>
  </div>
</body>
`;
